sphinx/environment.py

```diff
@@ -961,6 +962,29 @@ def process_downloads(self, docname, doctree):
 
     def process_images(self, docname, doctree):
         """Process and rewrite image URIs."""
+        def collect_candidates(imgpath, candidates):
+            globbed = {}
+            for filename in glob(imgpath):
+                new_imgpath = relative_path(path.join(self.srcdir, 'dummy'),
+                                            filename)
+                if filename.lower().endswith('.pdf'):
+                    if 'application/pdf' not in candidates:
+                        globbed.setdefault('application/pdf', []).append(new_imgpath)
```

Comment:
Well, I don't think we disagree then?
All the `if X not in candidates: globbed.setdefault(X, []).append(new_imgpath)` can be replaced with a simple `globbed.setdefault…` and further down, you'd have:

```python
for key, files in iteritems(globbed):
  if key not in candidates:
    candidates[key] = sorted(files, key=len)[0]  # select by similarity
```